src/Integration/Aws/SimpleS3/src/SimpleS3Client.php

```diff
@@ -47,6 +55,114 @@ public function has(string $bucket, string $key): bool
+     * } $options
+     */
+    public function copy(string $srcBucket, string $srcKey, string $destBucket, string $destKey, array $options = []): void
+    {
+        $megabyte = 1024 * 1024;
+        $sourceHead = $this->headObject(['Bucket' => $srcBucket, 'Key' => $srcKey]);
+        $contentLength = (int) $sourceHead->getContentLength();
+        $options['ContentType'] = $sourceHead->getContentType();
+        $concurrency = (int) ($options['Concurrency'] ?? 10);
+        $mupThreshold = ((int) ($options['mupThreshold'] ?? 2 * 1024)) * $megabyte;
+        unset($options['Concurrency'], $options['mupThreshold']);
+        /*
+         * The maximum number of parts is 10.000. The partSize must be a power of 2.
+         * We default this to 64MB per part. That means that we only support to copy
+         * files smaller than 64 * 10 000 = 640GB. If you are coping larger files,
+         * please set PartSize to a higher number, like 128, 256 or 512. (Max 4096).
+         */
+        $partSize = ($options['PartSize'] ?? 64) * $megabyte;
+        unset($options['PartSize']);
+
+        // If file is less than multipart upload threshold, use normal atomic copy
+        if ($contentLength < $mupThreshold) {
+            $this->copyObject(
+                CopyObjectRequest::create(
+                    array_merge($options, ['Bucket' => $destBucket, 'Key' => $destKey, 'CopySource' => "{$srcBucket}/{$srcKey}"])
+                )
+            );
+
+            return;
+        }
+
+        $uploadId = $this->createMultipartUpload(
+            CreateMultipartUploadRequest::create(
+                array_merge($options, ['Bucket' => $destBucket, 'Key' => $destKey])
+            )
+        )->getUploadId();
+        if (!$uploadId) {
+            throw new UnexpectedValue('UploadId can not be obtained');
+        }
+
+        $partNumber = 1;
+        $startByte = 0;
+        $parts = [];
+        while ($startByte < $contentLength) {
+            $parallelChunks = $concurrency;
+            /** @var UploadPartCopyOutput[] $responses */
+            $responses = [];
+            while ($startByte < $contentLength && $parallelChunks > 0) {
+                $endByte = min($startByte + $partSize, $contentLength) - 1;
+                $responses[$partNumber] = $this->uploadPartCopy(
+                    UploadPartCopyRequest::create([
+                        'Bucket' => $destBucket,
+                        'Key' => $destKey,
+                        'UploadId' => $uploadId,
+                        'CopySource' => "{$srcBucket}/{$srcKey}",
+                        'CopySourceRange' => "bytes={$startByte}-{$endByte}",
+                        'PartNumber' => $partNumber,
+                    ])
+                );
+
+                $startByte += $partSize;
+                ++$partNumber;
+                --$parallelChunks;
+            }
+            $error = null;
+            foreach ($responses as $idx => $response) {
```

Comment:
this is not efficient. We define 10 concurents upload, but in fact, we are uploading 10, waiting for the 10 to finish, and uploading 10 more, ...
That means:
- we are waiting for the 10th to finish before starting the 9 previous.
- if 1 upload takes time to finish, we don't leverage the 9 others.

A Beter implementation would be to use a pool (or buffer), to fill the pool with 10 responses, and once 1 response is over, add a new one before checking the other responses.

Better: leverage async processing, by not waiting for the end of a response, but checking if the response is over, if not check the next one.

```
while not all part uploaded
  while pool not full
    => start new upload
  while pool full
    foreach response in pull
      if response not finished
        continue
      else
        => process response
        => remove response from pool

foreach response in pull
  => wait for response to finish
  => process response
  => remove response from pool
```